### ISSUE TITLE
Blazor Pre-rendering and PersistingComponentStateSubscription not working correctly with dotnet 9?

### ISSUE THREAD

**Issue by MattCost** [dotnet/AspNetCore.Docs#34229]
### Description

Steps to reproduce
1. Make a new template `dotnet  new blazor`
1. add `@rendermode InteractiveServer` to the weather page
1. Start the app
1. Go to /weather
1 Observe the double rendering in action.
1. Add in code from https://learn.microsoft.com/en-us/aspnet/core/blazor/components/prerender?view=aspnetcore-9.0 to use PersistingComponentStateSubscription to save data
1. Restart the app
1. Navigate back and forth using the navigation links on the left hand side. (between home and weather and counter). Most times the double rendering still occurs on the weather page, sometimes it does not.
1. Navigate to /home via the link on the left.
1. Manually type in /weather into the url
1. Double rendering never occurs when manually typing in the url

Questions 

Why does the persisting only work sometimes?

The page has the following note `If the app adopts interactive routing and the page is reached via an internal navigation, prerendering doesn't occur. Therefore, you must perform a full page reload for the PrerenderedCounter2 component to see the following output.`

This doesn't make sense. Using internal navigation, prerendering is still occuring, and persisting is NOT working to stop it. Is this note out of date?

Are there changes in dotnet 9 to how the PersistingComponentStateSubscription works? or how pre-rendering works?

Not sure if this needs to be a bug report in another repo, or if this is just a documentation issue.

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/blazor/components/prerender?view=aspnetcore-9.0

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/components/prerender.md

### Document ID

ffdc8df3-55a8-b757-d4e6-38c104916002

### Article author

@guardrex

[Related Issues](https://github.com/dotnet/AspNetCore.Docs/issues?q=is%3Aissue+is%3Aopen+ffdc8df3-55a8-b757-d4e6-38c104916002)

**Comment by MattCost** [comment]
Sample repo https://github.com/MattCost/blazorTest

**Comment by guardrex** [comment]
Hello @MattCost ... 

I don't think the NOTE is incorrect; however, this behavior might be a bit more complex than what the NOTE states and require more explanation.

We're right on the doorstep of leaving for the 🦃 holiday, so I won't be able to get a product unit member to look at this today. I'll ping one of them next week to come on and take a look to see if we can improve the coverage.

**Comment by guardrex** [comment]
You ***might*** be able to get an answer today by copying this over to a new issue on the product unit's repo at ...

https://github.com/dotnet/aspnetcore/issues

***Leave this issue open tho. I think we'll be adding a bit more explanation, and I'll be working the doc update from this issue.*** This issue will close automatically later when the docs PR merges.

Please add ...

```
cc: @guardrex https://github.com/dotnet/AspNetCore.Docs/issues/34229
```

... to the bottom of your opening comment over there so that I can follow the discussion.

**Comment by MattCost** [comment]
Thanks for the quick response! Opened an issue over there in case there really is a bug.
Enjoy the long weekend!

**Comment by guardrex** [comment]
Sure thing ... I'll keep an 👁 on it to see where it goes.

**Comment by MattCost** [comment]
I think the docs def need to be updated somewhere. Enhanced Routing is the default, and enhanced routing doesn't work with Persisting Component State. That note isn't really clearly telling me it wont work.

**Comment by guardrex** [comment]
I think global interactive Server is the default, but VS remembers your last selection. Either way, it doesn't matter. This is challenging to cover because it is complicated and touches on several subjects. The interactive routing and enhanced nav coverage in the *Routing* article and the *Prerendering* coverage in the *Render Modes* article might need to be touched-up in addition to the coverage in this dedicated *Prerendering* article. When I get back next week and based on how your PU issue discussion unfolds, I'll take another look at all of this coverage to see what we can do to improve it. Here are the other places that I mentioned ...

* https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-9.0#interactive-routing
* https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-9.0#enhanced-navigation-and-form-handling
* https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-9.0#prerendering

**Comment by guardrex** [comment]
Also cross-referencing ... https://github.com/dotnet/aspnetcore/issues/51584

**Comment by MattCost** [comment]
>but VS remembers your last selection

I didn't use Visual Studio. I used the dotnet cli to make a new `blazor` app, and edited it with VSCode.  The template app chooses to omit the global interactive server attribute in the Routes.razor.

>This is challenging to cover because it is complicated and touches on several subjects

Exactly. I started with a default template project, added some sample code from the docs, and when it didn't work, stumbled into the nuances of how all the different blazor options interact with each other.

`If the app adopts interactive (enhanced) routing and the page is reached via an internal navigation, prerendering doesn't occur. `
I think this line is confusing. Ok, prerendering doesn't occur, but why did the page appear to flicker, and load the data twice?
Well according to this [helpful explaination](https://github.com/dotnet/aspnetcore/issues/59198#issuecomment-2504367245) from @javiercn the page is not getting rendered twice due to pre-rendering, but due to a quirk (?) of blazor.

```
To retain the initial value of the counter during prerendering, Blazor supports persisting state in a prerendered page using the [PersistentComponentState](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.persistentcomponentstate) service (and for components embedded into pages or views of Razor Pages or MVC apps, the [Persist Component State Tag Helper](https://learn.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/built-in/persist-component-state?view=aspnetcore-9.0)).
```

I think maybe just a disclaimer added above this section. Hey LISTEN up! If your app is configured with PerPage Interactivity, and uses Enhanced Navigation, then this PersistantComponentState is not passed between the server side and client side runtimes.

**Comment by YuriyDurov** [comment]
@MattCost I think that line is referring to the navigations happening under an interactive render mode.

You are observing this behavior with the NavMenu navigations, right? The default rendering behavior is that the app globally uses Static SSR and then you can enable interactivity for specific pages.

So what probably happens here is that, even if you enable interactivity on the page itself, the NavMenu is located outside of the scope of the page's interactivity. So the NavMenu behaves as if you are using Static SSR.

Correct me if I am wrong but I believe this could be an explanation to this. The Blazor circuits thing is hard to grasp comprehensively...

Could you try and check if the same behavior persists when the NavLink you are clicking is located inside the scope of the app's interactivity? For example, by enabling interactivity for the entire app or just for the NavMenu.

**Comment by MattCost** [comment]
>So what probably happens here is that, even if you enable interactivity on the page itself, the NavMenu is located outside of the scope of the page's interactivity. So the NavMenu behaves as if you are using Static SSR

Yes this is whats happening. I click the link, first there is a static server side render. Then there is a second interactive (client side?) render since that page is using InteractiveServer render mode. From a noob's perspective, I saw a double flash, and assumed it was just the pre-render / render behavior, and figured PersistComponentState would help (but it didn't). Hence this issue being created. 

You are correct. Setting the app to globally use InteractiveServer render mode makes everything behave as expected. Link is clicked, pre-render happens, state is persisted. Second render happens, data is fetched from state.

I'm satisfied with this explanation, just leaving this issue open for @guardrex  to work on the docs to explain some of this complex behavior somehow.

**Comment by guardrex** [comment]
**UPDATE (12/5)**: I'm still working my way back to this, but it's going to take a bit more time to reach it. There's internal stuff slowing things down ... holiday vacations ... exhausted devs 😩 (and *exhausted doc authors!* 🦖😆).

I'm working through backlogged issues, and I'll get back to this as soon as I can ... I hope within the next week or so 🤞🍀.

**Comment by guardrex** [comment]
The PR is up. This issue will close automatically after the PR merges.
